src/Simulation/Common/SimulatorBase.cs

```diff
@@ -21,6 +21,7 @@ public abstract class SimulatorBase : AbstractFactory<AbstractCallable>, IOperat
     {
         public event Action<ICallable, IApplyData> OnOperationStart = null;
         public event Action<ICallable, IApplyData> OnOperationEnd = null;
+        public event Action<System.Runtime.ExceptionServices.ExceptionDispatchInfo> OnFail = null;
```

Comment:
I hate to ask, but what is the `ExceptionDispatchInfo`, and do we really need to provide it?

Reply:
[`ExceptionDispatchInfo`](https://docs.microsoft.com/en-gb/dotnet/api/system.runtime.exceptionservices.exceptiondispatchinfo) contains exception itself and relevant debugging information from the moment exception was thrown. So if you say wanted to say access the call-stack of the exception at the moment when it was thrown, you could do it. Original exceptions is available through `ExceptionDispatchInfo.SourceException` property. 

From the docs:
> The ExceptionDispatchInfo object stores the stack trace information and Watson information that the exception contains at the point where it is captured. The exception can be thrown at another time and possibly on another thread by calling the ExceptionDispatchInfo.Throw method. The exception is thrown as if it had flowed from the point where it was captured to the point where the Throw method is called.

I thought it might be nice sometimes to know about the call-stack at the moment when exception was thrown.
But I am fine with just passing an exception itself.

Reply:
If you think that `ExceptionDispatchInfo` is an overkill, I could change the argument back to just `Exception`.